File: src/Enterprise/C#/DecoratorExample.cs
using System;

public interface IComponent
{
    string Render();
}

public sealed class PlainMessage : IComponent
{
    public string Render() => "alert";
}

public abstract class ComponentDecorator(IComponent inner) : IComponent
{
    protected IComponent Inner { get; } = inner;
    public abstract string Render();
}

public sealed class AuditDecorator(IComponent inner) : ComponentDecorator(inner)
{
    public override string Render() => $"audit({Inner.Render()})";
}

public sealed class EncryptDecorator(IComponent inner) : ComponentDecorator(inner)
{
    public override string Render() => $"enc({Inner.Render()})";
}

public static class DecoratorExample
{
    public static void Main()
    {
        IComponent component = new PlainMessage();
        Console.WriteLine($"base={component.Render()}");
        Console.WriteLine($"audit={new AuditDecorator(component).Render()}");
        Console.WriteLine($"encrypted={new EncryptDecorator(component).Render()}");
        Console.WriteLine($"stacked={new AuditDecorator(new EncryptDecorator(component)).Render()}");
    }
}

File: src/Enterprise/VisualBasic/DecoratorExample.vb
Imports System

Public Interface IComponent
    Function Render() As String
End Interface

Public NotInheritable Class PlainMessage
    Implements IComponent

    Public Function Render() As String Implements IComponent.Render
        Return "alert"
    End Function
End Class

Public MustInherit Class ComponentDecorator
    Implements IComponent

    Protected ReadOnly Inner As IComponent

    Protected Sub New(innerComponent As IComponent)
        Inner = innerComponent
    End Sub

    Public MustOverride Function Render() As String Implements IComponent.Render
End Class

Public NotInheritable Class AuditDecorator
    Inherits ComponentDecorator

    Public Sub New(innerComponent As IComponent)
        MyBase.New(innerComponent)
    End Sub

    Public Overrides Function Render() As String
        Return $"audit({Inner.Render()})"
    End Function
End Class

Public NotInheritable Class EncryptDecorator
    Inherits ComponentDecorator

    Public Sub New(innerComponent As IComponent)
        MyBase.New(innerComponent)
    End Sub

    Public Overrides Function Render() As String
        Return $"enc({Inner.Render()})"
    End Function
End Class

Module DecoratorExample
    Sub Main()
        Dim component As IComponent = New PlainMessage()
        Console.WriteLine($"base={component.Render()}")
        Console.WriteLine($"audit={New AuditDecorator(component).Render()}")
        Console.WriteLine($"encrypted={New EncryptDecorator(component).Render()}")
        Console.WriteLine($"stacked={New AuditDecorator(New EncryptDecorator(component)).Render()}")
    End Sub
End Module
